Feature: BB-164 support multiple connectors in oplog populator

welansari opened the pull request:
Issue: [BB-164](https://scality.atlassian.net/browse/BB-164)

Added support for the use of multiple connectors by the oplogPopulator

--------
**Connector Class**
This class manages the state of a kafka-connect mongodb source connectors.
The class contains the connector's config and can assign or unassign buckets from it

--------
**ConnectorsManager Class**
A ConnectorsManager keeps a list of active connectors

It is also used to spawn and configure connectors.

At startup, old connectors are first retreived and reconfigured*, then if more
connectors are required it creates them.

*Old connectors are reconfigured where buckets that are no longer valid
are removed from the connectors' config

--------
**Allocator Class**
The Allocator handles assigning buckets to connectors using
an allocation strategy.

--------
**Allocation Strategy Class**
Allocation strategy is what decides what connector should be assigned
to a bucket next

The allocation strategy implemented for now, returns the connector with the
fewest buckets assigned

--------
**OplogPopulator**
The OplogPopulator dynamically configures kafka-connect MongoDB source connectors to watch over changes in relevant buckets in MongoDB

At startup, the OplogPopulator first initializes connectors following its config.
If we have old connectors, these are reconfigured as explained above. More connectors are created if needed
The allocator is then instantiated, and used to start listening to valid buckets that don't have an assigned connector yet
A change stream is then established with the `__metastore` db to watch for changes of bucket configs, this is used to decide if we should listen to object operations in a bucket or not.
We start listening to a bucket if it uses at least one of the active backbeat extensions (ex. replication, notification...)

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/backbeat/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-164 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.1`


Please check the `Fix Version/s` of BB-164, or the target
branch of this pull request.

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -49,8 +49,12 @@ class OplogPopulator {
+        // stores connector assigned for each bucket
+        this._connectorToBucket = new Map();
+        // stores sets of buckets assigned to each connector
+        this._bucketsForConnectors = {};
```
should be a field inside each `Connector` (i.e. each entry of the `_configuredConnectors` list)

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -198,27 +202,57 @@ class OplogPopulator {
+            return assignedConnector;
+        } else {
+            // else return the connector with the fewest buckets assinged to it
+            const bucketPerConnector = this._configuredConnectors.map(connector =>
```
this algorithm will not scale (`O(N) * 3`), e.g. if we need to manage *many* connectors : we may need keep this list ordered somehow.

If the number of connectors stays low, this may be acceptable ; but at the moment we don't know how we will (need to) scale : e.g. if it will be better/needed to increase the number of connectors (and keep a low bucket count per connector) OR to increase the number of buckets per connector.

→ either we can make an easy algo upgrade (and be somewhat future-proof), or we need to create a TechDebt ticket (and address in BucketNotif or next EPIC)

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -198,27 +202,57 @@ class OplogPopulator {
-        this._bucketsForConnector.add(bucketName);
-        this.updateConnectorPipeline(constants.defaultConnectorName, [...this._bucketsForConnector]);
+    _listenToBucket(bucketName, updateConnector = true) {
+        const assignedConnector = this._assignConnectorForBucket(bucketName);
+        this._connectorToBucket.set(bucketName, assignedConnector);
```
the scope/role of the assignConnectorForBucket, and the followup "assignation", is not very consistent : either it should do indeed the assignation (and update the data), only simply handle the "retrieval" (and we must assign here if not already done)

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -198,27 +202,57 @@ class OplogPopulator {
+    /**
+     * start listening to a specific bucket's events
      * @param {string} bucketName bucket name
+     * @param {boolean} updateConnector should connector get updated
```
not clear this does (e.g. looking just at this function) : maybe this should not be part of this function, and/or the function prototype should be updated to better convey the intent of the function

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -391,25 +437,146 @@ class OplogPopulator {
+                config.pipeline = this._generateNewConnectorPipeline(
+                    [...this._bucketsForConnectors[connectorName]]
+                );
+                // creating or updating connector
```
nit: "create or update connector"

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -391,25 +437,146 @@ class OplogPopulator {
+     */
+    async _restoreConnectorBuckets(backbeatEnabledBuckets) {
+        try {
+            const activeConnectors = await this._connectWrapper.getConnectors();
```
`_connectWrapper` is not very representative: may be better to rename this to `_kafkaConnect`, so that code is easier to read/follow/understand

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -391,25 +437,146 @@ class OplogPopulator {
+                const bucketList = this._extractBucketsFromConfig(connectorConfig)
+                    .filter(bucket => backbeatEnabledBuckets.includes(bucket));
+                // storing the previous state
+                this._bucketsForConnectors[connectorName] = new Set([...bucketList]);
```
should split the code, I can see a few concerns (classes) we can isolate:
* the `Connector` itself (which has a number of buckets)
* the `allocation strategy` : not much for now, but mostly an interface with a contract, e.g. return the connector to use (or possibly create, in a future strategy : already now this could be used to delay the creation of connector until we need one) → will be easier to test if isolated
* the `Allocator` : e.g. handles the glue between the data (connectors) and (abstarct) allocation strategy → not sure if it should include the actual kafka-connect management, or move this elsewhere?

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -391,25 +437,146 @@ class OplogPopulator {
+                    .getConnectorConfig(connectorName);
+                // extracting buckets from config pipeline
+                const bucketList = this._extractBucketsFromConfig(connectorConfig)
+                    .filter(bucket => backbeatEnabledBuckets.includes(bucket));
```
this filtering is somewhat weird:
* need to actually "remove" the buckets from the connector, if this is not needed anymore
* i am concerned there may be a race condition : e.g. if a bucket is reconfigured after the initial list is read
* the scope of the function is somewhat confusing, seems to rely very strongly on the way it is called inside setup() instead of having a clear purpose (i.e. imagining it could be reused / checking if the function provides the "right" level of abstraction / more pragmatically: 1. can we review this function without knowing where it is called ; 2. can we then understand the calling code [setup] without looking at the imeplementation of this function)

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -391,25 +437,146 @@ class OplogPopulator {
+            // getting buckets with at least one backbeat extension active
+            const backbeatEnabledBuckets = await this._getBackbeatEnabledBuckets();
+            // saving previous state of connectors
+            await this._restoreConnectorBuckets(backbeatEnabledBuckets);
```
regarding "race condition", I guess the general logic here should be:

1. get the existing configuration from kafka-connect : e.g. see what is currently setup
2. do a "scan & watch" of mongo, to ensure we somewhat safely get the list of buckets to watch and all subsequent updates → update the connectors accordingly.

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -391,25 +437,146 @@ class OplogPopulator {
+            // creating/updating the connectors
+            await this._spawnConnectors();
+            // deleting unused connectors
+            await this._deleteUnusedConnectors();
```
not sure there is really a specific case here?
it could (should) happen after we stop listening to any bucket (though we may indeed want to have some mecanism to delay/bundle the check)

francoisferrand commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -391,25 +437,146 @@ class OplogPopulator {
+            // assigning a connector to bucket not in previous state
+            this._handleBucketsWithNoConnectors(backbeatEnabledBuckets);
+            // creating/updating the connectors
+            await this._spawnConnectors();
```
may be created on demand, which would provide more future-proof interface with the "allocation strategy" (i.e. if/when we want to support a strategy with max-number-of-buckets-per-connector)

francoisferrand reviewed (commented):
There is a corner case where we may on startup discover more connectors than the expected value in settings.

This is a somewhat valid use case (e.g. when "scaling down" the number of connectors), and The absolute invariant we must maintain is that a bucket must *never* be moved from one connector to another (since the ChangeStream offset is maintained by the connector) : thus we must not try to "reallocate" the bucket to another connector, and we must be able to live with that: e.g. the value in settings is a goal, not the "real" limit, and we may indeed have more connectors.

This also implies that extra connectors may empty-up afterwards (when un-listening to a bucket), get deleted later... which can lead to a "random" list of connector names (i.e. not just `[0-ConenctorsCount[`)

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-164 contains:


* `8.4.2`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.3`


Please check the `Fix Version/s` of BB-164, or the target
branch of this pull request.

Commit:
feature: BB-164 update OplogPopulatorTask

Commit:
feature: BB-164 add oplogPopulator Connector class

Connector class manages a kafka-connect mongodb source connector.
It contains its config and can assign or unassign buckets from it

Commit:
feature: BB-164 add OplogPopulator ConnectorsManager class

A ConnectorsManager keeps a list of active connectors

It is also used to spawn and configure connectors according to config.
Old connectors are first retreived and reconfigured, then if more
connectors are required it creates them.

Old connectors are reconfigured where buckets that are no longer valid
are removed from the connectors' config

Commit:
feature: BB-164 add OplogPopulator Allocator class

The Allocator handles assigning buckets to connectors using
an allocation strategy.

Commit:
feature: BB-164 add OplogPopulator allocation strategy

Allocation strategy is what decides what connector should be assigned
to a bucket next

The allocation strategy implemented for now, returns the connector with the
fewest buckets assigned

Commit:
feature: BB-164 add multiple connector support to OplogPopulator

Commit:
feature: BB-164 update OplogPopulator config to support multiple connectors

jbertran commented on extensions/oplogPopulator/modules/Connector.js:
```diff
@@ -0,0 +1,208 @@
+     */
+    getBucketCount() {
+        return this._buckets.size;
+    }
```
We can make these "standard" getters:

```js
get name() { return this.name; }    // get with instance.name
```

jbertran commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,278 @@
+        config['topic.namespace.map'] = JSON.stringify({
+            '*': this._oplogTopic,
+        });
+        return config;
```
In the spirit of avoiding side-effects:

```js
const connectorConfig = {
    name: connectorName,
    database: this._database,
    connection.uri: this._mongoUrl,
    topic.namespace.map: JSON.stringify({
        '*': this._oplogTopic,
    }),
}
return {
    ...constants.defaultConnectorConfig,
    ...connectorConfig
}
```

jbertran commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,278 @@
+     * @returns {string} generated connector name
+     */
+    _generateConnectorName() {
+        let connectorName = `${this._prefix}${constants.defaultConnectorName}`;
```
We have `uuid` installed in backbeat already, might be interesting instead of hand-rolling the naming algorithm.

jbertran commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,278 @@
+            // get connector config
+            const config = this._getDefaultConnectorConfiguration(connectorName);
+            // initialize connector
+            const connector = new Connector({
```
This introduces a time of check/time of use race condition on the connector name. If we use UUIDs, we largely avoid these (i.e. we stop checking because UUID is good enough). I'm leaning towaards deferring the name-_generating_ logic to the connector itself, and let the `ConnectorManager` decide whether or not it is valid.

jbertran commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,278 @@
+     * @returns {Promise|Connector} created connector
+     * @throws {InternalError}
+     */
+    async addConnector(connectors, spawn = true) {
```
We have access to `this._connectors` here, why are we not using it?

jbertran commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,278 @@
+     * Get currently active connectors
+     * @returns {Connectors[]} list of connectors
+     */
+    getConnectors() {
```
Ditto for `get`

```js
get connectors() {
    return this._connectors;
}
```

jbertran commented on extensions/oplogPopulator/modules/Allocator.js:
```diff
@@ -0,0 +1,124 @@
+     * @returns {Promise|undefined} undefined
+     * @throws {InternalError}
+     */
+    async listenToBucket(bucket, isretry = false) {
```
minor: `isRetry` for consistent snake-case

miniscruff replied:
That is not snake-case but camal casing. But I do agree.

jbertran commented on extensions/oplogPopulator/modules/Allocator.js:
```diff
@@ -0,0 +1,124 @@
+        joi.attempt(params, paramsJoi);
+        this._connectorsManager = params.connectorsManager;
+        this._logger = params.logger;
+        this._allocationStrategy = new LeastFullConnector({
```
So the connector list is frozen on startup? Is this expected?

welansari replied:
Good catch, it's not even used actually by the allocationStrategy
Just some old code that i forgot to remove while working on it

jbertran commented on extensions/oplogPopulator/allocationStrategy/LeastFullConnector.js:
```diff
@@ -0,0 +1,41 @@
+     * @returns {Connector} connector
+     */
+    getConnector(connectors) {
+        let candidat = null;
```
minor: `candidate`

jbertran commented on extensions/oplogPopulator/allocationStrategy/LeastFullConnector.js:
```diff
@@ -0,0 +1,41 @@
+     */
+    getConnector(connectors) {
+        let candidat = null;
+        connectors.forEach(connector => {
```
If you want to avoid side-effects, you can `reduce` instead:

```js
return connectors.reduce((prev, elt) => elt.getBucketCount() < prev.getBucketCount() ? elt : prev);
```

jbertran reviewed (changes requested):
Looks good, most of the comments are cosmetic

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-164 contains:


* `8.4.2`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.5`


Please check the `Fix Version/s` of BB-164, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers






The following reviewers are expecting changes from the author, or must review again:

* @jbertran

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

jonathan-gramain commented on extensions/oplogPopulator/modules/Connector.js:
```diff
@@ -0,0 +1,202 @@
+     * connector config
+     * @param {Object} params.buckets buckets assigned to this connector
+     * @param {Logger} params.logger logger object
+     * @param {Object} params.kafkaConnectHost kafka connect host
```
`{string}`?

jonathan-gramain commented on extensions/oplogPopulator/modules/Connector.js:
```diff
@@ -0,0 +1,202 @@
+     * @param {Object} params.buckets buckets assigned to this connector
+     * @param {Logger} params.logger logger object
+     * @param {Object} params.kafkaConnectHost kafka connect host
+     * @param {Object} params.kafkaConnectPort kafka connect port
```
`{number}`?

jonathan-gramain commented on extensions/oplogPopulator/modules/Connector.js:
```diff
@@ -0,0 +1,202 @@
+     * Remove bucket from this connector
+     * Connector is updated with new bucket list
+     * @param {string} bucket bucket to add
+     * @param {boolean} doUpdate updates connector if true
```
may use `[doUpdate=true]` syntax

jonathan-gramain commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,267 @@
+     * @param {string} params.mongoUrl MongoDB connection url
+     * @param {string} params.oplogTopic topic to use for oplog
+     * @param {string} params.kafkaConnectHost kafka connect host
+     * @param {string} params.kafkaConnectPort kafka connect port
```
`{number}`?

jonathan-gramain commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,267 @@
+            }
+            return connector;
+        } catch (err) {
+            // allowing one retry for when a connector name
```
in which case would a name be already taken? Apparently we use uuid so there should be virtually no chance of a conflict, hence no need to deal with this issue, unless I'm misreading it.

jonathan-gramain commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,267 @@
+            }));
+            return connectors;
+        } catch (err) {
+            this._logger.error('An error occured while getting old connectors', {
```
occured -> occurred (same for the other log messages around)

jonathan-gramain commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,267 @@
+                // extract buckets from old connector config and filter them
+                // only leaving currently valid buckets
+                const buckets = this._extractBucketsFromConfig(config);
+                // removing pipeline from config
```
why removing the pipeline? Also it could be better if we have to update the config, to do it in a more obvious place, like in the main function, to avoid unintended side-effects from a function starting with `_get` (that ideally should not have side effects), unless it would just be a cache for the get function but it does not look so.

jonathan-gramain commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,267 @@
+            // Add connectors if required number of connectors not reached
+            const nbConnectorsToAdd = this._nbConnectors - this._connectors.length;
+            if (nbConnectorsToAdd > 0) {
+                for (let i = 0; i < nbConnectorsToAdd; i++) {
```
I tend to prefer using `async` module functions such as `async.timesSeries`, it's more explicit and more flexible on what kind of concurrency we're getting, but I'm fine with how it is as a loop.

jonathan-gramain commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,267 @@
+            }
+            // Add connectors if required number of connectors not reached
+            const nbConnectorsToAdd = this._nbConnectors - this._connectors.length;
+            if (nbConnectorsToAdd > 0) {
```
This test looks redundant, could be removed

jonathan-gramain commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,267 @@
+            // getting connector's invalid buckets
+            const invalidBuckets = connector.buckets.filter(bucket =>
+                !buckets.includes(bucket));
+            if (invalidBuckets.length > 0) {
```
looks like we can remove this check, as `Promise.all` should run fine on an empty list

jonathan-gramain commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,267 @@
+                !buckets.includes(bucket));
+            if (invalidBuckets.length > 0) {
+                // removing invalid buckets
+                await Promise.all(invalidBuckets.map(bucket =>
```
How many invalid buckets could there be at worst? Asking to check if there could be so many (say more than 1000) that it might be a bad idea to do them all in parallel, for all connectors, and instead using `async.eachSeries`, `async.eachLimit` or equivalent could be better.

jonathan-gramain commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -0,0 +1,267 @@
+            this._logger.error('An error occured while removing invalid buckets from a connector', {
+                method: 'ConnectorsManager.removeConnectorInvalidBuckets',
+                error: err.description || err.message,
+                connector,
```
How is the connector going to be logged? I think it's better to just log its name for example, to avoid potentially a giant log line with a lot of internal fields logged.

jonathan-gramain commented on extensions/oplogPopulator/modules/Allocator.js:
```diff
@@ -0,0 +1,123 @@
+        } catch (err) {
+            // We allow one retry as in initial setup
+            // the connector might not be fully ready
+            if (!isRetry) {
```
One retry may not always be the best value, we can simply make it more flexible without extra complexity by passing `remainingRetries` instead, decrementing it at each call and stopping when reaching 0 (and then, choosing a reasonable initial value that could be 1 but also something else configurable).

Alternatively, `async.retry` provides support for such pattern and may also be tuned to do things like exponential wait times between retries to cope with transient issues - maybe not useful here but for future reference.

jonathan-gramain reviewed (commented):
Partial review, will continue tomorrow

jonathan-gramain commented on tests/unit/oplogPopulator/oplogPopulator.js:
```diff
@@ -629,202 +529,113 @@ describe('OplogPopulator', () => {
-            })
-            .catch(err => assert.ifError(err));
+    describe('_handleChangeStreamChangeEvent', () => {
+        it('should stop listening to bucket if got deleted', done => {
```
This series of tests may not require a `done` callback since they look to be synchronous.

welansari commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.4`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.4`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/8.3`



Please check the status of the associated issue BB-164.

Goodbye kerkesni.
